### PR TITLE
PID-3

### PR DESCRIPTION
**Pull request by jpflueger**
Implementation for PID-3. Some of the data is lost in conversion from FHIR->HL7 so I just took a best effort to try and re-create that data.

### CODE REVIEW

File: data/Templates/FhirToHL7v2/Mutators/_PatientIdentifiersToCX.liquid [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+{%endcomment%}
+{%- assign acceptedCodes = "MR, SS" | split: ", " -%}
+{%- capture acceptedSystemsStr -%}
+http://example.com/v2-to-fhir-converter/assigning-authority-universal-system-HCD,
+http://example.com/v2-to-fhir-converter/assigning-authority-local-system-USSSA,
+{%- endcapture -%}
+{%- assign acceptedSystems = acceptedSystemsStr | strip_newlines | split: "," -%}
+{%-comment-%} skip the first item that is sourced from PID-2 {%-endcomment-%}
+{%- assign filteredIds = identifiers | slice_array: 1 -%}
+{%- assign filteredIds = filteredIds | select_or: "type.coding[0].code", acceptedCodes -%}
```
Comment (jakiefermsft) [inline data/Templates/FhirToHL7v2/Mutators/_PatientIdentifiersToCX.liquid:11]: Do you want `coding[0]` here or `coding[]`? For me it's usually been the nonspecific index.

File: src/Microsoft.Health.Fhir.Liquid.Converter.FunctionalTests/TestData/Expected/FhirToHL7v2/ADT_A08/ADT-A08-01-expected.hl7 [hunk 2]
```diff
@@ -1,5 +1,5 @@
...
+PID|1||PATID1234^^^&b3ad3a9f-d4f6-4638-0a09-2da771733488&GUID^MR^^^^^^^~123456789^^^USSSA&&^SS^^^^^^^~PNM1234^^^&1fca4c8f-7562-b755-9a05-5f0f49096e5b&GUID^MR^^19241011^19241012^^^^||EVERYMAN&&&&^ADAM^A^III^Dr.^MD^D^^^^^19241012^^PF^Addsm~Josh&&&&^stanley^^^^^D^^^^^19241010^19241015^^|SMITH^Angela^L||M||2106-3^White^HL70005~2028-9^Asian^HL70005|1000 Hospital Lane^Ste. 123^Ann Arbor ^MI^99999^USA^^^^^^^20000110^20000120||2222^^^^^^^^^^^~12121212^^^^^^^^^^^~8484^^^^^^^^^^^~717171^^^^^^^^^^^|7777^^^^^^^^^^^~1111^^^^^^^^^^^~89898989^^^^^^^^^^^~021212^^^^^^^^^^^|ara^Arabic^HL70296^^^v2~eng^English^HL70296^^^v2.1|M^Married|AME|4000776^^^AccMgr&2.4&ISO^VN|||1212121^^^NTH&rt23&HCD^AND^^19241011^19241012
```
Comment (jakiefermsft) [inline src/Microsoft.Health.Fhir.Liquid.Converter.FunctionalTests/TestData/Expected/FhirToHL7v2/ADT_A08/ADT-A08-01-expected.hl7:4]: PID.3.4 here doesn't look right to me but I'm not sure.

Theirs:
`test1&2.16.1&HCD`

Yours:
`&b3ad3a9f-d4f6-4638-0a09-2da771733488&GUID`

File: src/Microsoft.Health.Fhir.Liquid.Converter/Filters/StringFilters.cs [hunk 3]
```diff
@@ -122,5 +123,10 @@ public static string RegexReplace(string data, string pattern, string replacemen
         {
             return Regex.Replace(data, pattern, replacement);
         }
+
+        public static bool StartsWith(string data, string delimiter)
+        {
+            return data.StartsWith(data, ignoreCase: true, culture: CultureInfo.InvariantCulture);
```
Comment (jakiefermsft) [inline src/Microsoft.Health.Fhir.Liquid.Converter/Filters/StringFilters.cs:129]: Is this relevant? https://docs.microsoft.com/en-us/dotnet/standard/base-types/best-practices-strings

> Do not use string operations based on StringComparison.InvariantCulture in most cases. One of the few exceptions is when you are persisting linguistically meaningful but culturally agnostic data.

File: src/Microsoft.Health.Fhir.Liquid.Converter.FunctionalTests/TestData/Expected/FhirToHL7v2/ADT_A08/ADT-A08-01-expected.hl7 (hunk 2 above)
Comment (jakiefermsft) [reply src/Microsoft.Health.Fhir.Liquid.Converter.FunctionalTests/TestData/Expected/FhirToHL7v2/ADT_A08/ADT-A08-01-expected.hl7:4]: Looking closer, it looks like you might need to search the patientOrganizations for those GUIDs. Those organization objects contain the rest of your values, like I showed yesterday with `patientMotherIdentifier`

File: data/Templates/FhirToHL7v2/Mutators/_PatientIdentifiersToCX.liquid [hunk 4]
```diff
@@ -0,0 +1,70 @@
+{%comment%}
+{%endcomment%}
+{%- assign acceptedCodes = "MR, SS" | split: ", " -%}
+{%- capture acceptedSystemsStr -%}
+http://example.com/v2-to-fhir-converter/assigning-authority-universal-system-HCD,
+http://example.com/v2-to-fhir-converter/assigning-authority-local-system-USSSA,
+{%- endcapture -%}
+{%- assign acceptedSystems = acceptedSystemsStr | strip_newlines | split: "," -%}
+{%-comment-%} skip the first item that is sourced from PID-2 {%-endcomment-%}
```
Comment (jakiefermsft) [inline data/Templates/FhirToHL7v2/Mutators/_PatientIdentifiersToCX.liquid:9]: I was curious if there might be another way to exclude this item. After looking at the object it seems there might be, through the organization, which leads me to wonder if we're going about this backwards.

I wonder if there's (we would probably know if we were actually solving a problem here) always "official" or "most authority" organizations that could help us locate the Identifiers/objects that are most "interesting" for PID.

🤔

File: data/Templates/FhirToHL7v2/Segments/_PID.liquid [hunk 5]
```diff
@@ -46,7 +48,7 @@
 {% comment %}  0 Message Type                          {% endcomment %}PID|
 {% comment %}  1 Set ID - PID                          {% endcomment %}1|{% comment %} PID.1 does not get sent to Fhir {% endcomment %}
 {% comment %}  2 Patient ID                            {% endcomment %}{% comment %}1234567^4^M11^test^MR^University Hospital^19241011^19241012{% endcomment %}|
-{% comment %}  3 Patient Identifier List               {% endcomment %}{% comment %}PATID1234^5^M11^test1&2.16.1&HCD^MR^GOOD HEALTH HOSPITAL~123456789^^^USSSA^SS{% endcomment %}|
```
Comment (jakiefermsft) [inline data/Templates/FhirToHL7v2/Segments/_PID.liquid:49]: As best I can tell, `GOOD HEALTH HOSPITAL` does not exist in the FHIR. I think it's helpful for reviewers if you can call this information out in the PR somewhere.

File: data/Templates/FhirToHL7v2/Mutators/_PatientIdentifiersToCX.liquid [hunk 6]
```diff
@@ -0,0 +1,70 @@
+{%comment%}
+{%endcomment%}
+{%- assign acceptedCodes = "MR, SS" | split: ", " -%}
+{%- capture acceptedSystemsStr -%}
+http://example.com/v2-to-fhir-converter/assigning-authority-universal-system-HCD,
```
[the remark this answers is not in the document]
Comment (jpflueger) [reply data/Templates/FhirToHL7v2/Mutators/_PatientIdentifiersToCX.liquid:5]: Ya I thought about adding a comparison operator or something but then it just feels like we're re-creating JMESPath implementation. Could maybe switch the call from string.equals with a call to a regex function?

File: data/Templates/FhirToHL7v2/Mutators/_PatientIdentifiersToCX.liquid (hunk 1 above)
Comment (jpflueger) [reply data/Templates/FhirToHL7v2/Mutators/_PatientIdentifiersToCX.liquid:11]: For all of the identifiers I've found there is only one coding value and I think in my reading of the fhir mapping that I just want the first one.

File: src/Microsoft.Health.Fhir.Liquid.Converter/Filters/StringFilters.cs (hunk 3 above)
Comment (jpflueger) [reply src/Microsoft.Health.Fhir.Liquid.Converter/Filters/StringFilters.cs:129]: Ya I thought I was being smart here but it looks my brain needs to be updated, will fix.

File: data/Templates/FhirToHL7v2/Mutators/_PatientIdentifiersToCX.liquid (hunk 4 above)
Comment (jpflueger) [reply data/Templates/FhirToHL7v2/Mutators/_PatientIdentifiersToCX.liquid:9]: I have some notes on this. I was trying to figure out the most succinct filter I could. This is the table of identifiers from the expected JSON file. 

| Name    | Value           | Code    | Assigner        | System |
| ------- | --------------- | ------- | --------------- | ------------ |
| PID-2   | 1234567         | MR      |                 | local/test |
| PID-3   | PATID1234       | MR      | Org/b3ad        | unvsl/HCD |
| PID-3   | 123456789       | SS      |                 | local/USSSA |
| PID-4   | PATID567        |         |                 | local/test2 |
| PID-19  | PSSN123121234   | SB      |                 | http://hl7.org/fhir/sid/us-ssn |
| PID-20  | DLN-123         | DL      |                 | http://example.com/v2-to-fhir-converter/CodeSystem/US |
| GT1-2   | 1516            | MR      |                 | local/test |
| GT1-19  | 55121           | FI      |                 |            |
| DB1-3   | DB123           | MR      | Org/4e68        | urn:oid:1.12 |
| IN2-6   | Medicare-12345  | MC      |                 | http://hl7.org/fhir/sid/us-medicare |
| IN2-8   | MCN-008         | MA      |                 | http://example.com/v2-to-fhir-converter/identifier-medicaid |
| IN2-10  | MI-12345        | MI      |                 |  |
| IN2-61  | PNM1234         | MR      | Org/1fca        | unvsl/HCD |

Which is why I landed on filtering by MR, SS and also HCD to try and eliminate GT1-2, DB1-3, etc.

File: data/Templates/FhirToHL7v2/Mutators/_PatientIdentifiersToCX.liquid (hunk 6 above)
[the remark this answers is not in the document]
Comment (jakiefermsft) [reply data/Templates/FhirToHL7v2/Mutators/_PatientIdentifiersToCX.liquid:5]: I tried JSONPath but it was insufficient. Can we replace ComplexObjectFilterUtility with JMESPath?

File: src/Microsoft.Health.Fhir.Liquid.Converter.FunctionalTests/TestData/Expected/FhirToHL7v2/ADT_A08/ADT-A08-01-expected.hl7 (hunk 2 above)
Comment (jakiefermsft) [reply src/Microsoft.Health.Fhir.Liquid.Converter.FunctionalTests/TestData/Expected/FhirToHL7v2/ADT_A08/ADT-A08-01-expected.hl7:4]: Nothing in the AZDO repo is relevant here.

File: data/Templates/FhirToHL7v2/Mutators/_PatientIdentifiersToCX.liquid (hunk 4 above)
Comment (jakiefermsft) [reply data/Templates/FhirToHL7v2/Mutators/_PatientIdentifiersToCX.liquid:9]: Then do we even need that slice to get rid of the first element? 🤔

File: src/Microsoft.Health.Fhir.Liquid.Converter.FunctionalTests/TestData/Expected/FhirToHL7v2/ADT_A08/ADT-A08-01-expected.hl7 (hunk 2 above)
Comment (jpflueger) [reply src/Microsoft.Health.Fhir.Liquid.Converter.FunctionalTests/TestData/Expected/FhirToHL7v2/ADT_A08/ADT-A08-01-expected.hl7]: Okay, it should match now but I don't think it's pretty.

**Review by jakiefermsft — APPROVED** [review]

**Comment by jakiefermsft** [comment]
Oh and just dropping this here so neither of us forget, don't forget to test this with the new test @jasonheld-msft added to main.

**Review by jasonheld-msft — APPROVED** [review]

**Pull request merged**